Optimize codegen for tuple swap scenarios

RikkiGibson opened the pull request:
Closes #53300

This PR optimizes scenarios like the following:

[SharpLab](https://sharplab.io/#v2:C4LglgNgPgAgTARgLAChUwCwAICyCAUYAdsFgB4A0WxpAngJSoDeqWbW+lWDWAvB7Spl6AblbsYCAJz5OVBqNQBfVEA=)
```cs
void M1(int x, int y)
{
    (x, y) = (y, x);
    Console.WriteLine((x, y));
}

```

Note that the compiler does additional work to preserve the order of evaluation for these assignments. For example, we have to take care that we read the 'x' on the RHS of the assignment before performing the store to 'x' on the LHS of the assignment. It's important that the semantics don't just end up being equivalent to `x = y; y = x;​` 🙂.

To accomplish this the compiler expands the above assignment into something like this:

[SharpLab](https://sharplab.io/#v2:C4LglgNgPgAgTARgLAChUwCwAICycAUYAdsFgB4A0WxpAngJSoDeqWb1JWwApgLYAOATSwBeLLQDcrdjS59+ADVHkpKduWU8Bg1etqb5C3exgIAnPnyVx9eqoC+QA===)
```cs
void M2(int x, int y)
{
    int tempY = y;
    int tempX = x;
    x = tempY;
    y = tempX;
    Console.WriteLine((x, y));
}

```

I was able to improve IL codegen for tuple assignment by simply reversing the order in which the temps are stored to the final variables. This enables us to preserve order of side-effects while also enabling the compiler's stack optimization pass to turn it into simpler "push and pop" operations.

[SharpLab](https://sharplab.io/#v2:C4LglgNgPgAgTARgLAChUwCwAICyBmACjADtgsAPAGixLIE8BKVAb1S3ZtK2AFMBbAA4BNLAF4sdANxYA9DKwCArgGcAFmw61u/AQA0xFaXIUr1KDhIO9Buo/IEB7ARvbkrOoXYVOXWGAgBOAgIqCQYGSVQAXyA=)
```cs
void M3(int x, int y)
{
    int tempY = y; // push
    int tempX = x; // push
    y = tempX; // pop
    x = tempY; // pop
    Console.WriteLine((x, y));
}
```

Using the above expansion, we are able to eliminate the locals 'tempY' and 'tempX' in IL, and end up with identical IL codegen to `int temp = y; y = x; x = temp​`.

There are plenty of scenarios where it would be breaking change to optimize in this way, such as `(x, x) = (a, b)`​. I've inserted some checks prevent the optimization from being used in cases like this, as well as when the LHS is an expression that could be side-effecting.

Here are the general requirements for performing the optimization:
1. the RHS is a tuple literal (which means the temps produced for this assignment are for the tuple elements, which could turn into push-pops into the destination variables)
2. at least one element in the RHS is actually stored to a temp. i.e. it is not a constant expression.
3. all variables on the LHS are unique, by-value, and are locals or parameters.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs:
```diff
@@ -103,6 +106,88 @@ internal sealed partial class LocalRewriter
+                        visitedSymbols.Add(expressionSymbol);
+                    }
+
+                    if (target.NestedVariables is { } nestedVariables && !canReorderTargetAssignments(nestedVariables, ref visitedSymbols))
```
>if (target.NestedVariables [](http://example.com/codeflow?start=20&length=26)

Is there a good reason in performing this check if ```if (target.Single is { } single)``` succeeded above?

Also, we checked ```target.Single``` and ```target.NestedVariables```. Is there an alternative state that ```target``` can have? If not, then consider structuring the code so that it is clear that these are the only states we can run into.

RikkiGibson replied:
Have corrected this to expect exactly one of Single or NestedVariables to be present.

AlekseyTs replied:
>> Is there an alternative state that target can have? If not, then consider structuring the code so that it is clear that these are the only states we can run into.  

> Have corrected this to expect exactly one of Single or NestedVariables to be present.

The code is written as though the ```target.NestedVariables is { } nestedVariables``` test can fail at this point. If that is not the case, then we can simply apply an unconditional cast in place of the first argument for the following ```canReorderTargetAssignments``` call. If the test can fail, then it looks like we not handling this situation and assuming that reordering can be done for it.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs:
```diff
@@ -103,6 +106,88 @@ internal sealed partial class LocalRewriter
+            void reverseAssignmentsToTargetsIfApplicable()
+            {
+                // The list of deconstruction targets is usually small. Prefer a temporary array which is stack-bound for small numbers of elements.
+                var visitedSymbols = TemporaryArray<Symbol>.Empty;
```
>TemporaryArray<Symbol>.Empty; [](http://example.com/codeflow?start=37&length=29)

I would use pooled hashset or a SmallDictionary since we are looking for members in the array

RikkiGibson replied:
I'm open to changing this, but will hold off until we have decided more about how we want this optimization to work. For example, do we want to break behavior of any possibly aliasing assignments in order to get better codegen in some scenarios we care about.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs:
```diff
@@ -103,6 +106,88 @@ internal sealed partial class LocalRewriter
+                            return false;
+                        }
+
+                        var expressionSymbol = single.ExpressionSymbol;
```
>ExpressionSymbol [](http://example.com/codeflow?start=54&length=16)

I am very uncomfortable with usage of  ```ExpressionSymbol``` API outside of SemanticModel. It was added to support SemanticModel and is implemented with that in mind. We are enumerating all interesting nodes above and can easily extract relevant symbols from them with a very small refactoring.  Also, there is no point to add/lookup anything for BoundDiscardExpression.

RikkiGibson replied:
Have corrected this. Thanks

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs:
```diff
@@ -103,6 +106,88 @@ internal sealed partial class LocalRewriter
+                        {
+                            if ((object)visitedSymbol == expressionSymbol)
+                            {
+                                // This deconstruction writes to the same target multiple times, e.g:
```
>// This deconstruction writes to the same target multiple times, e.g: [](http://example.com/codeflow?start=32&length=69)

It is not clear why do we care if that is the case.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs:
```diff
@@ -103,6 +106,88 @@ internal sealed partial class LocalRewriter
+                var visitedSymbols = TemporaryArray<Symbol>.Empty;
+
+                if (right is BoundConvertedTupleLiteral or BoundConversion { Operand.Kind: BoundKind.TupleLiteral or BoundKind.ConvertedTupleLiteral }
+                    && effects.init.Any()
```
>effects.init.Any() [](http://example.com/codeflow?start=23&length=18)

It would be good to capture what we are trying to infer with this check. I assume we are detecting presence of temp assignments that are supposed to precede assignments to the targets. Correct?

RikkiGibson replied:
Yes, I'll try and make that more clear.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs:
```diff
@@ -103,6 +106,88 @@ internal sealed partial class LocalRewriter
+                // The list of deconstruction targets is usually small. Prefer a temporary array which is stack-bound for small numbers of elements.
+                var visitedSymbols = TemporaryArray<Symbol>.Empty;
+
+                if (right is BoundConvertedTupleLiteral or BoundConversion { Operand.Kind: BoundKind.TupleLiteral or BoundKind.ConvertedTupleLiteral }
```
>BoundKind.TupleLiteral or BoundKind.ConvertedTupleLiteral [](http://example.com/codeflow?start=91&length=57)

Do tests cover both options?

RikkiGibson replied:
Looked at this a little closer. I think testing for BoundKind.TupleLiteral is unnecessary. Removing that subpattern and adding an assertion that this is never the kind of the RHS.

AlekseyTs commented:
Done with review pass (commit 3)

RikkiGibson commented:
Thank you for the feedback. I have addressed most of the comments. I think we need a little more time to decide the exact scope of the optimization, but the new changes are still reviewable.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs:
```diff
@@ -103,6 +106,102 @@ internal sealed partial class LocalRewriter
+                // The list of deconstruction targets is usually small. Prefer a temporary array which is stack-bound for small numbers of elements.
+                var visitedSymbols = TemporaryArray<Symbol>.Empty;
+
+                Debug.Assert(right is not BoundConversion { Operand.Kind: BoundKind.TupleLiteral });
```
>right is not BoundConversion { Operand.Kind: BoundKind.TupleLiteral } [](http://example.com/codeflow?start=29&length=69)

Consider also asserting that ```right``` is not any form of a tuple literal itself.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs:
```diff
@@ -103,6 +106,102 @@ internal sealed partial class LocalRewriter
+                    Debug.Assert(target is { Single: not null, NestedVariables: null } or { Single: null, NestedVariables: not null });
+                    if (target.Single is { } single)
+                    {
+                        Symbol? symbol = single switch
```
>Symbol? symbol = single switch [](http://example.com/codeflow?start=24&length=30)

I understand that there is a desire to use new language features, but, in my opinion, a switch statement can incorporate logic from this statement and from the ```if``` below in a much clearer way. The logic won't be spread out across multiple conditional constructs, and it might actually be not much longer in terms of lines used. Up to you.

AlekseyTs commented:
> I think we need a little more time to decide the exact scope of the optimization, but the new changes are still reviewable.

It might take some time for us to make a decision on whether we would like to change behavior for other scenarios. I would be comfortable with handling more scenarios separately, if we decide to do so later.

RikkiGibson commented:
> It might take some time for us to make a decision on whether we would like to change behavior for other scenarios. I would be comfortable with handling more scenarios separately, if we decide to do so later.

Yeah, let's just do the optimization outlined in this PR, which we think we can do without any breaking changes. We'll see if we decide to expand on it a future PR.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs:
```diff
@@ -103,6 +105,96 @@ internal sealed partial class LocalRewriter
+            {
+                PooledHashSet<Symbol>? visitedSymbols = null;
+
+                Debug.Assert(right is not ({ Kind: BoundKind.TupleLiteral } or BoundConversion { Operand.Kind: BoundKind.TupleLiteral }));
```
>BoundKind.TupleLiteral [](http://example.com/codeflow?start=51&length=22)

``` or BoundKind.ConvertedTupleLiteral```?

RikkiGibson replied:
I think we do expect that `right.Kind` can be `ConvertedTupleLiteral` here.

AlekseyTs reviewed (approved):
LGTM (commit 9)

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs:
```diff
@@ -103,6 +105,96 @@ internal sealed partial class LocalRewriter
+            }
+
+            static bool canReorderTargetAssignments(ArrayBuilder<Binder.DeconstructionVariable> targets, ref PooledHashSet<Symbol>? visitedSymbols)
+            {
```
nit: Thanks for writing up the summary in OP. Consider capturing it here as comment:
```
Here are the general requirements for performing the optimization:
- the RHS is a tuple literal (which means the temps produced for this assignment are for the tuple elements, which could turn into push-pops into the destination variables)
- at least one element in the RHS is actually stored to a temp. i.e. it is not a constant expression.
- all variables on the LHS are unique, by-value, and are locals or parameters.
```

RikkiGibson replied:
Have adjusted the comments to include these, inline with the checks which correspond to these requirements.

jcouv reviewed (approved):
LGTM Thanks (iteration 9)
